src/common/dav_v2/heap.c

```diff
@@ -2126,6 +2274,50 @@ heap_get_zone_limits(uint64_t heap_size, uint64_t cache_size, uint32_t nemb_pct)
 	return zd;
 }
 
+int
+heap_incr_empty_nemb_cnt(struct palloc_heap *heap)
+{
+	return ++heap->rt->empty_nemb_cnt;
+}
+
+int
+heap_decr_empty_nemb_cnt(struct palloc_heap *heap)
+{
+	return heap->rt->empty_nemb_cnt ? --heap->rt->empty_nemb_cnt : 0;
+}
+
+int
+heap_force_recycle(struct palloc_heap *heap)
+{
+	struct bucket *defb;
+	struct mbrt   *mb;
+	uint32_t       zone_id;
+
+	if (!heap->rt->empty_nemb_gcth) {
+		heap->rt->empty_nemb_gcth = HEAP_NEMB_EMPTY_THRESHOLD;
+		d_getenv_uint("DAOS_NEMB_EMPTY_RECYCLE_THRESHOLD", &heap->rt->empty_nemb_gcth);
+	}
+
+	if (heap->rt->empty_nemb_cnt < heap->rt->empty_nemb_gcth)
+		return 0;
+
+	mb   = heap_mbrt_get_mb(heap, 0);
+	defb = mbrt_bucket_acquire(mb, DEFAULT_ALLOC_CLASS_ID);
+	while (heap_reclaim_next_ne(heap, &zone_id) == 0) {
+		heap_reclaim_zone_garbage(heap, defb, zone_id);
+		heap_reclaim_setlast_ne(heap, zone_id);
+	}
+	heap_reclaim_garbage(heap, defb);
+	mbrt_bucket_release(defb);
+	heap_populate_nemb_unused(heap);
+
+	if (heap->rt->empty_nemb_cnt >= heap->rt->empty_nemb_gcth)
+		D_ERROR("Force GC failed to free up enough nembs, cnt = %d",
+			heap->rt->empty_nemb_cnt);
+
+	return 0;
```

Comment:
Could you estimate the CPU time & transaction size in extreme case (assume meta blob size is 1TB, most of the zones are empty and to be reclaimed case).

If it could generate too large tx or hog CPU for too long (> 5ms), I think we need to introduce a limit for each round of force recycle.

Reply:
I have changed the logic in heap_force_recycle(). It will not try doing garbage reclaim in an incremental fashion everytime it gets called rather than wait for the number of empty buckets to reach threshold. With this each call to heap_force_recycle() is taking roughly 0.5us to 7us.